hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ServerRemoteProcedure.java

```diff
@@ -80,17 +82,25 @@ public abstract class ServerRemoteProcedure extends Procedure<MasterProcedureEnv
   protected ServerName targetServer;
   protected boolean dispatched;
   protected boolean succ;
+  // after remoteProcedureDone we require error field to decide the next state
+  protected Throwable remoteError;
+  protected MasterProcedureProtos.ServerRemoteProcedureState state =
+    MasterProcedureProtos.ServerRemoteProcedureState.SERVER_REMOTE_PROCEDURE_DISPATCH;
 
   protected abstract void complete(MasterProcedureEnv env, Throwable error);
 
   @Override
   protected synchronized Procedure<MasterProcedureEnv>[] execute(MasterProcedureEnv env)
     throws ProcedureYieldException, ProcedureSuspendedException, InterruptedException {
-    if (dispatched) {
+    if (
+      state != MasterProcedureProtos.ServerRemoteProcedureState.SERVER_REMOTE_PROCEDURE_DISPATCH
+    ) {
+      complete(env, this.remoteError);
       if (succ) {
```

Comment:
So here we expect the sub classes to set the succ flag? Can we just determine this by ourselves? If remoteError is null, then we are succeeded?

Reply:
Ping @Umeshkumar9414 .

Reply:
> Can we just determine this by ourselves? If remoteError is null, then we are succeeded
For the majority, this is the case but for some procedures like SnapshotVerifyProcedure (It also sets succ true in case of CorruptedSnapshotException {a DoNotRetryIOException}) and SplitWALRemoteProcedure (it also sets succ true in case of DoNotRetryIOException) have their logic to decide the succ flag. We can try to generalize these cases, but we need to think about other DoNotRetryIOException cases as well.

Reply:
I checked that it is only used in the `complete` method so we can make it as a return from the `complete` method. (Thought of renaming the `complete` also occurs but I don't have strong suggestions )

Reply:
I think making complete method return a boolean is a possible solution. Let's give it a try.

Reply:
It is done and now I think there is no need for succ flag. same as dispatched flag also that is also replaced by state.